node/src/components/consensus/highway_core/active_validator.rs

```diff
@@ -295,22 +302,34 @@ impl<C: Context> ActiveValidator<C> {
     /// Returns a new unit with the given data, and the correct sequence number.
     fn new_unit(
         &mut self,
-        mut panorama: Panorama<C>,
+        panorama: Panorama<C>,
         timestamp: Timestamp,
         value: Option<C::ConsensusValue>,
         state: &State<C>,
         instance_id: C::InstanceId,
         rng: &mut NodeRng,
-    ) -> SignedWireUnit<C> {
+    ) -> Option<SignedWireUnit<C>> {
         if let Some((prop_time, _)) = self.next_proposal.take() {
             warn!(
                 ?timestamp,
                 "canceling proposal for {} due to unit", prop_time
             );
         }
+        for hash in panorama.iter_correct_hashes() {
+            if timestamp < state.unit(hash).timestamp {
+                error!(
+                    %timestamp, justification_timestamp = %state.unit(hash).timestamp,
+                    "canceling unit creation because of outdated timestamp"
+                );
+                return None;
+            }
+        }
```

Comment:
Since both `timestamp` and `panorama` are passed into this method, why not make the check on the caller side? I am not suggesting doing that, I am just wondering which choice is better.

Reply:
I _think_ we are checking this on the caller side, that's why I made it an `error!`…
I'm hesitating to making it a `panic!`/`assert_ge!`… :grimacing: